src/main/scala/cromwell/engine/workflow/ValidateActor.scala

```diff
@@ -38,44 +49,88 @@ class ValidateActor(wdlSource: WdlSource, workflowInputs: Option[WdlJson], workf
   override def receive = {
     case ValidateWorkflow =>
       validateWorkflow(sender())
-      // NOTE: self shuts down when the parent PerRequest shuts down
-  }
-
-  private def validateInputs(namespaceWithWorkflow: NamespaceWithWorkflow, maybeWorkflowInputs: Option[WdlJson]): Future[Unit] = {
-    maybeWorkflowInputs match {
-      case Some(wi) =>
-        for {
-          inputs <- Future(wi.parseJson).map(_.asJsObject.fields)
-          coercedInputs <- Future.fromTry(namespaceWithWorkflow.coerceRawInputs(inputs))
-        } yield ()
-      case None => Future.successful(())
-    }
+    // NOTE: self shuts down when the parent PerRequest shuts down
   }
 
   private def validateWorkflow(sentBy: ActorRef): Unit = {
     logger.info(s"$tag for $sentBy")
-    val futureValidation: Future[Unit] = for {
+    val futureValidation: Future[ValidationResult] = (for {
       namespaceWithWorkflow <- Future(NamespaceWithWorkflow.load(wdlSource))
-      validatedInputs <- validateInputs(namespaceWithWorkflow, workflowInputs)
-      validatedRuntimeOptions <- Future.fromTry(TryUtil.sequence(namespaceWithWorkflow.workflow.calls map {
-        call => CromwellRuntimeAttributes.validateKeys(call.task.runtimeAttributes.attrs.keySet, CromwellBackend.backend.backendType)
-      }))
-    } yield ()
+      validatedInputs: Option[WorkflowCoercedInputs] <- Future(validateInputs(namespaceWithWorkflow))
+      validatedWorkflowOptions: Option[WorkflowOptions] <- Future(validateWorkflowOptions(CromwellBackend.backend()))
+      validatedRuntimeAttrs: Seq[Set[String]] <- Future(validateRuntimeAttributes(namespaceWithWorkflow))
+    } yield ValidationSuccess(namespaceWithWorkflow, validatedInputs, validatedWorkflowOptions, validatedRuntimeAttrs)) recover {
+      case reason: Throwable => ValidationFailure(reason)
+    }
+
+    //TODO: IMO, the responsibility for returning a status code should be a part of the client. This class should only return a validation result,
+    // and the callee should convert that to a status code or anything that it wants
 
     // Now validate that this Future completed successfully:
-    futureValidation onComplete {
-      case Success(_) =>
+    futureValidation map {
```

Comment:
I believe that will make the code harder to read. This Future is guaranteed to succeed (see `recover`), as such the onComplete will always yield a success, which then you'll have to pattern match against `Success(ValidationFailure)` (besides the compiler complaining about potential match errors because of a missing `case Failure(..)` , a line which will never be reached even if added.